File: examples/mobile/android/CodexTaskClient.kt
package xin.wangye.codextask

import android.util.Base64
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject

data class PromptDocument(val name: String, val content: String)
data class RemoteImage(val name: String, val mimeType: String, val bytes: ByteArray)
data class JobReceipt(val jobId: String, val statusUrl: String)

class CodexTaskClient(
    baseUrl: String,
    private val token: String,
    private val http: OkHttpClient = OkHttpClient(),
) {
    private val baseUrl = baseUrl.trimEnd('/')
    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()
    private val terminal = setOf("completed", "needs_input", "failed", "cancelled")

    suspend fun submitText(
        prompt: String,
        promptFiles: List<PromptDocument> = emptyList(),
        images: List<RemoteImage> = emptyList(),
        schema: JSONObject? = null,
    ): JobReceipt = submit("/v1/text", payload(prompt, promptFiles, images).apply {
        put("backend", "direct")
        put("reasoning", "medium")
        if (schema != null) put("schema", schema)
    })

    suspend fun submitImage(
        prompt: String,
        promptFiles: List<PromptDocument> = emptyList(),
        images: List<RemoteImage> = emptyList(),
        quality: String = "high",
    ): JobReceipt = submit("/v1/image", payload(prompt, promptFiles, images).apply {
        put("backend", "direct")
        put("quality", quality)
        put("count", 1)
    })

    suspend fun submitTask(
        prompt: String,
        workingDirectory: String,
        promptFiles: List<PromptDocument> = emptyList(),
        images: List<RemoteImage> = emptyList(),
    ): JobReceipt = submit("/v1/task", payload(prompt, promptFiles, images).apply {
        put("workingDirectory", workingDirectory)
        put("sandboxMode", "danger-full-access")
        put("networkAccess", true)
    })

    suspend fun resume(
        taskId: String,
        answer: String,
        promptFiles: List<PromptDocument> = emptyList(),
        images: List<RemoteImage> = emptyList(),
    ): JobReceipt = submit("/v1/tasks/$taskId/resume", payload(answer, promptFiles, images))

    suspend fun awaitJob(receipt: JobReceipt, pollEveryMillis: Long = 1_000): JSONObject {
        while (true) {
            val snapshot = getJson(receipt.statusUrl)
            if (snapshot.getString("status") in terminal) return snapshot
            delay(pollEveryMillis)
        }
    }

    suspend fun downloadArtifact(downloadUrl: String): ByteArray = withContext(Dispatchers.IO) {
        val request = authorized(Request.Builder().url(resolve(downloadUrl))).get().build()
        http.newCall(request).execute().use { response ->
            if (!response.isSuccessful) error("Artifact download failed: HTTP ${response.code}")
            response.body?.bytes() ?: error("Artifact response is empty")
        }
    }

    private suspend fun submit(path: String, body: JSONObject): JobReceipt {
        val json = postJson(path, body)
        return JobReceipt(json.getString("jobId"), json.getString("statusUrl"))
    }

    private suspend fun postJson(path: String, body: JSONObject): JSONObject = withContext(Dispatchers.IO) {
        val request = authorized(Request.Builder().url(resolve(path)))
            .post(body.toString().toRequestBody(jsonMediaType))
            .build()
        executeJson(request)
    }

    private suspend fun getJson(path: String): JSONObject = withContext(Dispatchers.IO) {
        executeJson(authorized(Request.Builder().url(resolve(path))).get().build())
    }

    private fun executeJson(request: Request): JSONObject {
        http.newCall(request).execute().use { response ->
            val text = response.body?.string().orEmpty()
            if (!response.isSuccessful) error("CodexTask HTTP ${response.code}: $text")
            return JSONObject(text)
        }
    }

    private fun payload(
        prompt: String,
        promptFiles: List<PromptDocument>,
        images: List<RemoteImage>,
    ): JSONObject = JSONObject().apply {
        put("prompt", prompt)
        put("promptFiles", JSONArray(promptFiles.map { JSONObject().put("name", it.name).put("content", it.content) }))
        put("images", JSONArray(images.map {
            JSONObject()
                .put("name", it.name)
                .put("mimeType", it.mimeType)
                .put("dataBase64", Base64.encodeToString(it.bytes, Base64.NO_WRAP))
        }))
    }

    private fun authorized(builder: Request.Builder): Request.Builder =
        builder.header("Authorization", "Bearer $token")

    private fun resolve(path: String): String = if (path.startsWith("http")) path else "$baseUrl$path"
}

File: examples/mobile/ios/CodexTaskClient.swift
import Foundation

struct PromptDocument {
    let name: String
    let content: String
}

struct RemoteImage {
    let name: String
    let mimeType: String
    let data: Data
}

struct JobReceipt {
    let jobId: String
    let statusURL: String
}

enum CodexTaskClientError: Error {
    case invalidResponse
    case http(Int, String)
    case server(String)
}

final class CodexTaskClient {
    private let baseURL: URL
    private let token: String
    private let session: URLSession
    private let terminalStatuses = Set(["completed", "needs_input", "failed", "cancelled"])

    init(baseURL: URL, token: String, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.token = token
        self.session = session
    }

    func submitText(
        prompt: String,
        promptFiles: [PromptDocument] = [],
        images: [RemoteImage] = [],
        schema: [String: Any]? = nil
    ) async throws -> JobReceipt {
        var body = payload(prompt: prompt, promptFiles: promptFiles, images: images)
        body["backend"] = "direct"
        body["reasoning"] = "medium"
        if let schema { body["schema"] = schema }
        return try await submit(path: "/v1/text", body: body)
    }

    func submitImage(
        prompt: String,
        promptFiles: [PromptDocument] = [],
        images: [RemoteImage] = []
    ) async throws -> JobReceipt {
        var body = payload(prompt: prompt, promptFiles: promptFiles, images: images)
        body["backend"] = "direct"
        body["quality"] = "high"
        body["count"] = 1
        return try await submit(path: "/v1/image", body: body)
    }

    func submitTask(
        prompt: String,
        workingDirectory: String,
        promptFiles: [PromptDocument] = [],
        images: [RemoteImage] = []
    ) async throws -> JobReceipt {
        var body = payload(prompt: prompt, promptFiles: promptFiles, images: images)
        body["workingDirectory"] = workingDirectory
        body["sandboxMode"] = "danger-full-access"
        body["networkAccess"] = true
        return try await submit(path: "/v1/task", body: body)
    }

    func resume(
        taskId: String,
        answer: String,
        promptFiles: [PromptDocument] = [],
        images: [RemoteImage] = []
    ) async throws -> JobReceipt {
        try await submit(
            path: "/v1/tasks/\(taskId)/resume",
            body: payload(prompt: answer, promptFiles: promptFiles, images: images)
        )
    }

    func awaitJob(_ receipt: JobReceipt, pollEveryNanoseconds: UInt64 = 1_000_000_000) async throws -> [String: Any] {
        while true {
            let snapshot = try await getJSON(path: receipt.statusURL)
            if let status = snapshot["status"] as? String, terminalStatuses.contains(status) { return snapshot }
            try await Task.sleep(nanoseconds: pollEveryNanoseconds)
        }
    }

    func downloadArtifact(path: String) async throws -> Data {
        var request = URLRequest(url: try resolve(path))
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        let (data, response) = try await session.data(for: request)
        try validate(response: response, data: data)
        return data
    }

    private func submit(path: String, body: [String: Any]) async throws -> JobReceipt {
        let json = try await postJSON(path: path, body: body)
        guard let jobId = json["jobId"] as? String, let statusURL = json["statusUrl"] as? String else {
            throw CodexTaskClientError.invalidResponse
        }
        return JobReceipt(jobId: jobId, statusURL: statusURL)
    }

    private func postJSON(path: String, body: [String: Any]) async throws -> [String: Any] {
        var request = URLRequest(url: try resolve(path))
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        return try await executeJSON(request)
    }

    private func getJSON(path: String) async throws -> [String: Any] {
        var request = URLRequest(url: try resolve(path))
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        return try await executeJSON(request)
    }

    private func executeJSON(_ request: URLRequest) async throws -> [String: Any] {
        let (data, response) = try await session.data(for: request)
        try validate(response: response, data: data)
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw CodexTaskClientError.invalidResponse
        }
        return json
    }

    private func validate(response: URLResponse, data: Data) throws {
        guard let http = response as? HTTPURLResponse else { throw CodexTaskClientError.invalidResponse }
        guard (200..<300).contains(http.statusCode) else {
            throw CodexTaskClientError.http(http.statusCode, String(data: data, encoding: .utf8) ?? "")
        }
    }

    private func payload(
        prompt: String,
        promptFiles: [PromptDocument],
        images: [RemoteImage]
    ) -> [String: Any] {
        [
            "prompt": prompt,
            "promptFiles": promptFiles.map { ["name": $0.name, "content": $0.content] },
            "images": images.map {
                ["name": $0.name, "mimeType": $0.mimeType, "dataBase64": $0.data.base64EncodedString()]
            },
        ]
    }

    private func resolve(_ path: String) throws -> URL {
        if let absolute = URL(string: path), absolute.scheme != nil { return absolute }
        guard let resolved = URL(string: path, relativeTo: baseURL)?.absoluteURL else {
            throw CodexTaskClientError.server("Invalid URL: \(path)")
        }
        return resolved
    }
}
